numpy/core/src/multiarray/mapping.c

```diff
@@ -1783,35 +1780,35 @@ array_assign_subscript(PyArrayObject *self, PyObject *ind, PyObject *op)
         return -1;
     }
 
-    /* Single field access */
-    if (PyDataType_HASFIELDS(PyArray_DESCR(self))) {
-        if (PyString_Check(ind) || PyUnicode_Check(ind)) {
-            PyObject *obj;
+    /* field access */
+    if (PyDataType_HASFIELDS(PyArray_DESCR(self)) &&
+            obj_is_string_or_stringlist(ind)) {
+        PyObject *obj;
+        static PyObject *indexfunc = NULL;
 
-            obj = PyDict_GetItem(PyArray_DESCR(self)->fields, ind);
-            if (obj != NULL) {
-                PyArray_Descr *descr;
-                int offset;
-                PyObject *title;
+        if (!(PyString_Check(ind) || PyUnicode_Check(ind))) {
```

Comment:
Also, there are already lots of unit tests for Py3k and unicode, that's how I found out I needed to be more careful. These lines happened to already be safe because they are only reached after going through `obj_is_string_or_stringlist`, but it is clearer/better if I add checks too.